File: macclean/cleaners/ios_backups.py
import plistlib
import shutil
from pathlib import Path
import click
from rich.console import Console
from rich.table import Table
from rich.panel import Panel
from macclean.core.utils import CleanItem, AnalysisResult, format_size, confirm, dir_size

console = Console()


def _read_backup_name(backup_dir: Path) -> str:
    info_plist = backup_dir / "Info.plist"
    if info_plist.exists():
        try:
            with open(info_plist, "rb") as f:
                data = plistlib.load(f)
            device = data.get("Device Name", "")
            product = data.get("Product Type", "")
            date = data.get("Last Backup Date", "")
            date_str = str(date)[:10] if date else "unknown date"
            return f"{device} ({product}) — {date_str}"
        except Exception:
            pass
    return backup_dir.name


def analyze(home: Path | None = None) -> AnalysisResult:
    home = home or Path.home()
    result = AnalysisResult()
    backup_root = home / "Library" / "Application Support" / "MobileSync" / "Backup"
    if not backup_root.exists():
        return result

    for backup_dir in backup_root.iterdir():
        if not backup_dir.is_dir():
            continue
        name = _read_backup_name(backup_dir)
        size = dir_size(backup_dir)
        result.items.append(CleanItem(label=name, path=backup_dir, size_bytes=size))

    return result


def clean(result: AnalysisResult, dry_run: bool = False, yes: bool = False) -> None:
    if not result.items:
        console.print("[green]No iOS backups found.[/]")
        return

    table = Table(show_header=True, show_lines=True)
    table.add_column("Device")
    table.add_column("Size", justify="right")
    for item in result.items:
        table.add_row(item.label, format_size(item.size_bytes))

    console.print(Panel(table, title="[bold cyan]iOS/iPadOS Backups[/]"))
    console.print(f"  Total recoverable: [bold]{format_size(result.total_bytes)}[/]")
    console.print("  [dim]Warning: deleting removes the only local backup for that device.[/]")

    if dry_run:
        return
    if not yes and not confirm("Delete all listed backups?"):
        return

    for item in result.items:
        try:
            shutil.rmtree(item.path, ignore_errors=True)
            console.print(f"  [green]✓[/] Deleted backup: {item.label}")
        except Exception as e:
            console.print(f"  [yellow]⚠[/] {item.label}: {e}")


@click.command()
@click.pass_context
def cmd(ctx):
    """List and delete iPhone/iPad local backups."""
    result = analyze()
    clean(result, dry_run=ctx.obj["dry_run"], yes=ctx.obj["yes"])

File: src/cleaners/ios_backups.rs
use crate::core::fs::dir_size;
use crate::core::{AnalysisResult, CleanKind, Cleaner, RiskLevel};
use crate::ui;
use anyhow::Result;
use std::path::PathBuf;

pub struct IosBackupsCleaner;

impl Cleaner for IosBackupsCleaner {
    fn name(&self) -> &str {
        "ios-backups"
    }
    fn display_name(&self) -> &str {
        "iOS Backups"
    }

    fn analyze(&self) -> Result<AnalysisResult> {
        let home = dirs::home_dir().unwrap_or_else(|| PathBuf::from("/tmp"));
        let mut result = AnalysisResult::default();

        let backup_root = home.join("Library/Application Support/MobileSync/Backup");
        if !backup_root.exists() {
            return Ok(result);
        }

        let entries = match std::fs::read_dir(&backup_root) {
            Ok(e) => e,
            Err(_) => return Ok(result),
        };

        for entry in entries.filter_map(|e| e.ok()) {
            let path = entry.path();
            if !path.is_dir() {
                continue;
            }

            // Try to get a friendly name from Info.plist; fall back to dir name
            let label = try_read_device_name(&path).unwrap_or_else(|| {
                path.file_name()
                    .map(|n| n.to_string_lossy().to_string())
                    .unwrap_or_else(|| "Unknown Backup".to_string())
            });

            let size = dir_size(&path);
            result.add_with_meta(
                label,
                path,
                size,
                CleanKind::Backup,
                RiskLevel::High,
                "Local iPhone/iPad backup copy; deleting can remove the only local restore point.",
            );
        }

        Ok(result)
    }

    fn clean(&self, result: &AnalysisResult, dry_run: bool, yes: bool) -> Result<()> {
        if result.items.is_empty() {
            println!("No iOS backups found.");
            return Ok(());
        }
        ui::print_analysis("iOS Backups", &result.items);
        ui::print_warn("Deleting removes only the local backup copy.");
        if dry_run {
            return Ok(());
        }
        if !yes && !ui::confirm("Delete all listed backups?", false)? {
            return Ok(());
        }

        for (path, outcome) in crate::core::trash::trash_clean_items("ios-backups", &result.items) {
            match outcome {
                Ok(_) => ui::print_ok(&format!("Moved backup to Trash: {}", path.display())),
                Err(e) => ui::print_warn(&format!("{}: {}", path.display(), e)),
            }
        }
        Ok(())
    }
}

/// Attempt to read the device name from Info.plist (XML format only).
/// Returns None if the plist is binary or the key is not found.
fn try_read_device_name(backup_dir: &std::path::Path) -> Option<String> {
    let plist_path = backup_dir.join("Info.plist");
    let contents = std::fs::read_to_string(&plist_path).ok()?;

    // Only handle XML plists (binary starts with "bplist")
    if contents.starts_with("bplist") {
        return None;
    }

    // Look for <key>Device Name</key> followed by <string>...</string>
    let key_marker = "<key>Device Name</key>";
    let key_pos = contents.find(key_marker)?;
    let after_key = &contents[key_pos + key_marker.len()..];

    let string_start = after_key.find("<string>")? + "<string>".len();
    let string_end = after_key[string_start..].find("</string>")?;
    let device_name = after_key[string_start..string_start + string_end]
        .trim()
        .to_string();

    if device_name.is_empty() {
        None
    } else {
        Some(device_name)
    }
}
